File: backend/src/main/java/be/lutske/leolegacy/interfaceadapter/rest/RecipeImportResource.java
package be.lutske.leolegacy.interfaceadapter.rest;

import be.lutske.leolegacy.application.service.ExtractionResult;
import be.lutske.leolegacy.application.service.RecipeExtractionException;
import be.lutske.leolegacy.application.service.RecipeExtractionService;
import be.lutske.leolegacy.infrastructure.persistence.entity.RecipeEntity;
import be.lutske.leolegacy.infrastructure.persistence.repository.CategoryRepository;
import be.lutske.leolegacy.infrastructure.persistence.repository.RecipeRepository;
import be.lutske.leolegacy.interfaceadapter.rest.RecipeImportDtos.ImportConfirmRequest;
import be.lutske.leolegacy.interfaceadapter.rest.RecipeImportDtos.ImportExtractionResponse;
import be.lutske.leolegacy.interfaceadapter.rest.RecipeImportDtos.ProposedRecipeDto;
import jakarta.transaction.Transactional;
import jakarta.ws.rs.BadRequestException;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import org.jboss.resteasy.reactive.RestForm;
import org.jboss.resteasy.reactive.multipart.FileUpload;

import java.io.IOException;
import java.nio.file.Files;
import java.time.Instant;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * REST resource for importing recipes from images via AI-powered extraction.
 *
 * <p>Flow:
 * <ol>
 *   <li>{@code POST /api/recipes/import} — always returns HTTP 200 with extraction for user review</li>
 *   <li>{@code POST /api/recipes/import/confirm} — saves the recipe after user review/correction</li>
 * </ol>
 */
@Path("/api/recipes/import")
public class RecipeImportResource {

    private static final long MAX_FILE_SIZE = 10L * 1024 * 1024; // 10 MB
    private static final long DEFAULT_IMPORT_CATEGORY_ID = 15L; // "Imported"
    private static final Set<String> ALLOWED_CONTENT_TYPES = Set.of(
            "image/png", "image/jpeg", "image/jpg", "image/webp"
    );
    private static final Set<String> ALLOWED_EXTENSIONS = Set.of("png", "jpg", "jpeg", "webp");

    private final RecipeExtractionService extractionService;
    private final RecipeRepository recipeRepository;
    private final CategoryRepository categoryRepository;

    public RecipeImportResource(RecipeExtractionService extractionService,
                                RecipeRepository recipeRepository,
                                CategoryRepository categoryRepository) {
        this.extractionService = extractionService;
        this.recipeRepository = recipeRepository;
        this.categoryRepository = categoryRepository;
    }

    /**
     * Upload an image and extract a recipe via AI (multimodal LLM).
     * Always returns HTTP 200 with the extraction result for user review.
     * The recipe is NOT saved — use {@code /confirm} to save after review.
     */
    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.APPLICATION_JSON)
    public Response importFromImage(@RestForm("file") FileUpload file) {
        validateFile(file);

        byte[] imageBytes;
        try {
            imageBytes = Files.readAllBytes(file.uploadedFile());
        } catch (IOException e) {
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
                    .entity(Map.of("error", "Failed to read uploaded file: " + e.getMessage()))
                    .build();
        }

        String mimeType = file.contentType() != null ? file.contentType() : "image/png";

        ExtractionResult extraction;
        try {
            extraction = extractionService.extractRecipeFromImage(imageBytes, mimeType);
        } catch (RecipeExtractionException e) {
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
                    .entity(Map.of("error", "AI extraction failed: " + e.getMessage()))
                    .build();
        }

        var proposed = toProposedDto(extraction);
        String status = extraction.isComplete() ? "COMPLETE" : "NEEDS_MORE_INFO";

        return Response.ok(new ImportExtractionResponse(
                status,
                extraction.rawModelResponse(),
                proposed,
                extraction.missingFields(),
                extraction.warnings()
        )).build();
    }

    /**
     * Confirm and finalize an imported recipe after user review.
     */
    @POST
    @Path("/confirm")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    @Transactional
    public Response confirmImport(ImportConfirmRequest request) {
        var proposed = request.proposedRecipe();
        var overrides = request.userOverrides();

        String title = override(overrides != null ? overrides.title() : null, proposed.title());
        List<String> ingredients = override(overrides != null ? overrides.ingredients() : null, proposed.ingredients());
        String preparation = override(overrides != null ? overrides.preparation() : null, proposed.preparation());
        Long categoryId = override(overrides != null ? overrides.categoryId() : null, proposed.categoryId());
        String notes = overrides != null ? overrides.notes() : proposed.notes();

        if (title == null || title.isBlank()) {
            throw new BadRequestException("Title is required");
        }
        if (ingredients == null || ingredients.isEmpty()) {
            throw new BadRequestException("Ingredients are required");
        }
        if (preparation == null || preparation.isBlank()) {
            throw new BadRequestException("Preparation is required");
        }

        String finalPreparation = (notes != null && !notes.isBlank())
                ? preparation + "\n\nNotes: " + notes
                : preparation;

        var merged = new ProposedRecipeDto(
                title, proposed.description(), proposed.servings(),
                ingredients, finalPreparation, proposed.source(), proposed.tags(),
                categoryId != null ? categoryId : DEFAULT_IMPORT_CATEGORY_ID,
                null
        );

        var recipe = createRecipeFromProposal(merged, request.rawModelResponse());
        return Response.status(Response.Status.CREATED)
                .entity(toDetailResponse(recipe))
                .build();
    }

    private <T> T override(T overrideValue, T originalValue) {
        return overrideValue != null ? overrideValue : originalValue;
    }

    private void validateFile(FileUpload file) {
        long fileSize = file.uploadedFile().toFile().length();
        if (fileSize > MAX_FILE_SIZE) {
            throw new BadRequestException("File too large: " + (fileSize / 1024 / 1024) + " MB. Maximum is 10 MB.");
        }

        String contentType = file.contentType();
        if (contentType != null && !ALLOWED_CONTENT_TYPES.contains(contentType)) {
            throw new BadRequestException("Unsupported file type: " + contentType + ". Allowed: PNG, JPG, JPEG, WEBP.");
        }

        String fileName = file.fileName();
        String extension = fileName.contains(".")
                ? fileName.substring(fileName.lastIndexOf('.') + 1).toLowerCase()
                : "";
        if (!ALLOWED_EXTENSIONS.contains(extension)) {
            throw new BadRequestException("Unsupported file extension: ." + extension + ". Allowed: .png, .jpg, .jpeg, .webp.");
        }
    }

    /**
     * Convert ExtractionResult to ProposedRecipeDto.
     * Steps (list) are joined into a single preparation string to match the frontend contract.
     */
    private ProposedRecipeDto toProposedDto(ExtractionResult extraction) {
        String preparation = extraction.steps() != null
                ? String.join("\n", extraction.steps())
                : null;

        return new ProposedRecipeDto(
                extraction.title(),
                extraction.description(),
                extraction.servings(),
                extraction.ingredients(),
                preparation,
                extraction.source(),
                extraction.tags(),
                null, // categoryId — set on confirm
                null  // notes
        );
    }

    private RecipeEntity createRecipeFromProposal(ProposedRecipeDto proposal, String rawModelResponse) {
        long categoryId = proposal.categoryId() != null ? proposal.categoryId() : DEFAULT_IMPORT_CATEGORY_ID;
        var category = categoryRepository.findById(categoryId);
        if (category == null) {
            category = categoryRepository.findById(DEFAULT_IMPORT_CATEGORY_ID);
        }
        if (category == null) {
            throw new IllegalStateException("Default import category not found");
        }

        var recipe = new RecipeEntity();
        recipe.setTitle(proposal.title() != null ? proposal.title() : "Untitled Recipe");
        recipe.setIngredients(proposal.ingredients() != null ? String.join("|", proposal.ingredients()) : "");
        recipe.setPreparation(proposal.preparation() != null ? proposal.preparation() : "");
        recipe.setCategory(category);
        recipe.setViewCount(0);
        recipe.setSource(proposal.source() != null ? proposal.source() : "Imported from image");
        recipe.setCreatedAt(Instant.now());

        if (rawModelResponse != null) {
            recipe.setImportMetadata(rawModelResponse);
        }

        recipeRepository.persist(recipe);
        return recipe;
    }

    private RecipeDetailResponse toDetailResponse(RecipeEntity entity) {
        List<String> ingredientList = Arrays.stream(entity.getIngredients().split("\\|"))
                .filter(s -> !s.isBlank())
                .toList();
        return new RecipeDetailResponse(
                entity.getId(),
                entity.getTitle(),
                ingredientList,
                entity.getPreparation(),
                entity.getCategory().getId(),
                entity.getCategory().getName(),
                entity.getViewCount()
        );
    }
}

File: backend/src/main/kotlin/be/lutske/leolegacy/interfaceadapter/rest/RecipeImportResource.kt
package be.lutske.leolegacy.interfaceadapter.rest

import be.lutske.leolegacy.application.service.ExtractionResult
import be.lutske.leolegacy.application.service.RecipeExtractionException
import be.lutske.leolegacy.application.service.RecipeExtractionService
import be.lutske.leolegacy.infrastructure.persistence.entity.RecipeEntity
import be.lutske.leolegacy.infrastructure.persistence.repository.CategoryRepository
import be.lutske.leolegacy.infrastructure.persistence.repository.RecipeRepository
import jakarta.transaction.Transactional
import jakarta.ws.rs.BadRequestException
import jakarta.ws.rs.Consumes
import jakarta.ws.rs.POST
import jakarta.ws.rs.Path
import jakarta.ws.rs.Produces
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import org.jboss.resteasy.reactive.multipart.FileUpload
import org.jboss.resteasy.reactive.RestForm
import java.time.Instant

/**
 * REST resource for importing recipes from images via LLM-based extraction.
 *
 * The image is sent to a multimodal LLM (OpenAI, Claude, or vLLM) through
 * LangChain4j. The model extracts structured recipe data which is validated
 * and either stored directly or returned for user correction.
 */
@Path("/api/recipes/import")
class RecipeImportResource(
    private val extractionService: RecipeExtractionService,
    private val recipeRepository: RecipeRepository,
    private val categoryRepository: CategoryRepository
) {

    companion object {
        private const val MAX_FILE_SIZE = 10L * 1024 * 1024 // 10 MB
        private const val DEFAULT_IMPORT_CATEGORY_ID = 15L // "Geimporteerd"
        private val ALLOWED_CONTENT_TYPES = setOf(
            "image/png", "image/jpeg", "image/jpg", "image/webp"
        )
        private val ALLOWED_EXTENSIONS = setOf("png", "jpg", "jpeg", "webp")
    }

    /**
     * Upload an image and attempt to extract a recipe via LLM.
     *
     * Returns:
     * - 201 Created with RecipeDetailResponse if recipe was fully extracted and saved
     * - 422 Unprocessable Entity with ImportNeedsMoreInfoResponse if extraction is incomplete
     * - 400 Bad Request if file validation fails
     * - 502 Bad Gateway if the LLM provider is unavailable
     */
    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Produces(MediaType.APPLICATION_JSON)
    @Transactional
    fun importFromImage(@RestForm("file") file: FileUpload): Response {
        // Validate file
        validateFile(file)

        // Read file bytes and determine MIME type
        val imageBytes = file.uploadedFile().toFile().readBytes()
        val mimeType = file.contentType() ?: "image/jpeg"

        // Extract recipe via LLM
        val extraction: ExtractionResult = try {
            extractionService.extractRecipeFromImage(imageBytes, mimeType)
        } catch (e: RecipeExtractionException) {
            return Response.status(502)
                .entity(mapOf("error" to "AI extraction failed: ${e.message}"))
                .build()
        }

        // If all required fields are present, create the recipe directly
        if (extraction.isComplete()) {
            val proposal = extractionToProposal(extraction)
            val recipe = createRecipeFromProposal(proposal, extraction)
            return Response.status(Response.Status.CREATED)
                .entity(recipe.toDetailResponse())
                .build()
        }

        // Otherwise, return needs-more-info
        return Response.status(422)
            .entity(
                ImportNeedsMoreInfoResponse(
                    rawModelResponse = extraction.rawModelResponse,
                    proposedRecipe = extractionToProposal(extraction),
                    missingFields = extraction.missingFields(),
                    warnings = extraction.warnings
                )
            )
            .build()
    }

    /**
     * Confirm and finalize an imported recipe after user provides missing info.
     */
    @POST
    @Path("/confirm")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    @Transactional
    fun confirmImport(request: ImportConfirmRequest): Response {
        val proposed = request.proposedRecipe
        val overrides = request.userOverrides

        // Merge proposed + overrides
        val title = overrides?.title ?: proposed.title
            ?: throw BadRequestException("Title is required")
        val ingredients = overrides?.ingredients ?: proposed.ingredients
            ?: throw BadRequestException("Ingredients are required")
        val preparation = overrides?.preparation ?: proposed.preparation
            ?: throw BadRequestException("Preparation is required")
        val categoryId = overrides?.categoryId ?: proposed.categoryId ?: DEFAULT_IMPORT_CATEGORY_ID
        val notes = overrides?.notes ?: proposed.notes

        if (ingredients.isEmpty()) {
            throw BadRequestException("At least one ingredient is required")
        }

        val mergedProposal = ProposedRecipeDto(
            title = title,
            ingredients = ingredients,
            preparation = if (notes.isNullOrBlank()) preparation else "$preparation\n\nNotities: $notes",
            categoryId = categoryId,
            description = overrides?.description ?: proposed.description,
            servings = overrides?.servings ?: proposed.servings,
            source = proposed.source
        )

        val recipe = createRecipeFromProposal(mergedProposal, null)
        return Response.status(Response.Status.CREATED)
            .entity(recipe.toDetailResponse())
            .build()
    }

    private fun validateFile(file: FileUpload) {
        // Check file size
        val fileSize = file.uploadedFile().toFile().length()
        if (fileSize > MAX_FILE_SIZE) {
            throw BadRequestException("File too large: ${fileSize / 1024 / 1024} MB. Maximum is 10 MB.")
        }

        // Check content type
        val contentType = file.contentType()
        if (contentType != null && contentType !in ALLOWED_CONTENT_TYPES) {
            throw BadRequestException("Unsupported file type: $contentType. Allowed: PNG, JPG, JPEG, WEBP.")
        }

        // Check file extension
        val fileName = file.fileName()
        val extension = fileName.substringAfterLast('.', "").lowercase()
        if (extension !in ALLOWED_EXTENSIONS) {
            throw BadRequestException("Unsupported file extension: .$extension. Allowed: .png, .jpg, .jpeg, .webp.")
        }
    }

    /**
     * Convert an [ExtractionResult] to a [ProposedRecipeDto].
     */
    private fun extractionToProposal(extraction: ExtractionResult): ProposedRecipeDto {
        return ProposedRecipeDto(
            title = extraction.title,
            description = extraction.description,
            servings = extraction.servings,
            ingredients = extraction.ingredients,
            preparation = extraction.steps?.joinToString("\n"),
            source = extraction.source,
            tags = extraction.tags
        )
    }

    private fun createRecipeFromProposal(proposal: ProposedRecipeDto, extraction: ExtractionResult?): RecipeEntity {
        val categoryId = proposal.categoryId ?: DEFAULT_IMPORT_CATEGORY_ID
        val category = categoryRepository.findById(categoryId)
            ?: categoryRepository.findById(DEFAULT_IMPORT_CATEGORY_ID)
            ?: throw IllegalStateException("Default import category not found")

        val recipe = RecipeEntity()
        recipe.title = proposal.title ?: "Naamloos recept"
        recipe.ingredients = proposal.ingredients?.joinToString("|") ?: ""
        recipe.preparation = proposal.preparation ?: ""
        recipe.category = category
        recipe.viewCount = 0
        recipe.source = proposal.source ?: "Imported from image"
        recipe.createdAt = Instant.now()

        // Store extraction metadata as JSON if available
        if (extraction != null) {
            val metadata = buildString {
                append("{")
                append("\"provider\":\"${extraction.provider ?: "unknown"}\",")
                append("\"model\":\"${extraction.model ?: "unknown"}\",")
                append("\"warnings\":")
                append(extraction.warnings.joinToString(",", "[", "]") { "\"${it.replace("\"", "\\\"")}\"" })
                append("}")
            }
            recipe.importMetadata = metadata
        }

        recipeRepository.persist(recipe)
        return recipe
    }

    private fun RecipeEntity.toDetailResponse() = RecipeDetailResponse(
        id = this.id,
        title = this.title,
        ingredients = this.ingredients.split("|").filter { it.isNotBlank() },
        preparation = this.preparation,
        categoryId = this.category.id,
        categoryName = this.category.name,
        viewCount = this.viewCount
    )
}
